domains/client/domain-executor/src/domain_block_processor.rs

```diff
@@ -182,6 +196,130 @@ where
 
+    pub(crate) fn fork_aware_block_imports(
+        &self,
+        primary_hash: PBlock::Hash,
+        primary_number: NumberFor<PBlock>,
+    ) -> sp_blockchain::Result<ForkAwareBlockImports<Block, PBlock>> {
+        if primary_number == One::one() {
+            return Ok(ForkAwareBlockImports {
+                initial_parent: (self.client.info().genesis_hash, Zero::zero()),
+                primary_imports: vec![HashAndNumber {
+                    hash: primary_hash,
+                    number: primary_number,
+                }],
+            });
+        }
+
+        let best_hash = self.client.info().best_hash;
+        let best_number = self.client.info().best_number;
+        let best_receipt = crate::aux_schema::load_execution_receipt::<
+            _,
+            Block::Hash,
+            NumberFor<PBlock>,
+            PBlock::Hash,
+        >(&*self.client, best_hash)?
+        .ok_or_else(|| {
+            sp_blockchain::Error::Backend(format!("Execution receipt not found for {best_hash:?}"))
+        })?;
+
+        let primary_from = best_receipt.primary_hash;
+        let primary_to = primary_hash;
+
+        if primary_from == primary_to {
+            return Err(sp_blockchain::Error::Application(Box::from(
+                "Primary block {primary_hash:?} has already been processed.",
+            )));
+        }
+
+        let route =
+            sp_blockchain::tree_route(&*self.primary_chain_client, primary_from, primary_to)?;
+
+        let retracted = route.retracted();
+        let enacted = route.enacted();
+
+        tracing::trace!(
+            ?retracted,
+            ?enacted,
+            common_block = ?route.common_block(),
+            "Calculating ForkAwareBlockImports on #{best_number},{best_hash:?}"
+        );
+
+        match (retracted.is_empty(), enacted.is_empty()) {
+            (true, false) => {
+                // New tip, A -> B
+                Ok(ForkAwareBlockImports {
+                    initial_parent: (self.client.info().best_hash, self.client.info().best_number),
+                    primary_imports: enacted.to_vec(),
+                })
+            }
+            (false, true) => {
+                let common_block_number = translate_number_type(route.common_block().number);
+                let parent_header = self
+                    .client
+                    .header(BlockId::Number(common_block_number))?
+                    .ok_or_else(|| {
+                        sp_blockchain::Error::Backend(format!("Header not found for {best_hash:?}"))
+                    })?;
+
+                Ok(ForkAwareBlockImports {
+                    initial_parent: (parent_header.hash(), *parent_header.number()),
+                    primary_imports: retracted.iter().cloned().rev().collect::<Vec<_>>(),
+                })
+            }
+            (true, true) => {
+                unreachable!("Tree route must not be empty as `from` and `to` in tree_route() must not be the same.");
```

Comment:
This looks more like a proof:
```suggestion
                unreachable!("Tree route is not empty because `primary_from` and `primary_to` in tree_route() are not be the same.");
```